dora/core/server/worker/src/main/java/alluxio/worker/s3/S3HttpHandler.java

```diff
@@ -0,0 +1,108 @@
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * Handles S3 HTTP request.
+ */
+public class S3HttpHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+  private static final Logger LOG = LoggerFactory.getLogger(S3HttpHandler.class);
+
+  private final FileSystem mFileSystem;
+
+  private final DoraWorker mDoraWorker;
+
+  private final AsyncUserAccessAuditLogWriter mAsyncAuditLogWriter;
+
+  private final ThreadPoolExecutor mLightPool;
+
+  private final ThreadPoolExecutor mHeavyPool;
+
+  /**
+   * Constructs an instance of {@link S3HttpHandler}.
+   *
+   * @param fileSystem
+   * @param doraWorker
+   * @param asyncAuditLogWriter
+   * @param lightPool
+   * @param heavyPool
+   */
+  public S3HttpHandler(FileSystem fileSystem, DoraWorker doraWorker,
+                       AsyncUserAccessAuditLogWriter asyncAuditLogWriter,
+                       ThreadPoolExecutor lightPool, ThreadPoolExecutor heavyPool) {
+    mFileSystem = fileSystem;
+    mDoraWorker = doraWorker;
+    mAsyncAuditLogWriter = asyncAuditLogWriter;
+    mLightPool = lightPool;
+    mHeavyPool = heavyPool;
+  }
+
+  //  private final PagedService mPagedService;
+  @Override
+  public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+    ctx.flush();
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext context, FullHttpRequest request)
+      throws Exception {
+    try {
+      S3NettyHandler s3Handler =
+          S3NettyHandler.createHandler(context, request, mFileSystem, mDoraWorker,
+              mAsyncAuditLogWriter);
+      // Handle request async
+      if (Configuration.getBoolean(PropertyKey.WORKER_S3_ASYNC_PROCESS_ENABLED)) {
+        S3NettyBaseTask.OpTag opTag = s3Handler.getS3Task().mOPType.getOpTag();
+        ExecutorService es = (ExecutorService) (opTag == S3NettyBaseTask.OpTag.LIGHT
+            ? mLightPool : mHeavyPool);
+
+        es.submit(() -> {
+          try {
+            HttpResponse response = s3Handler.getS3Task().continueTask();
+            s3Handler.processHttpResponse(response);
+          } catch (Throwable th) {
+            HttpResponse errorResponse =
+                S3ErrorResponse.createNettyErrorResponse(th, request.uri());
+            ChannelFuture f = context.writeAndFlush(errorResponse);
+            f.addListener(ChannelFutureListener.CLOSE);
+          }
+        });
+      } else {
+        HttpResponse response = s3Handler.getS3Task().continueTask();
+        s3Handler.processHttpResponse(response);
+      }
+    } catch (Exception ex) {
+      HttpResponse errorResponse = S3ErrorResponse.createNettyErrorResponse(ex, request.uri());
+      ChannelFuture f = context.writeAndFlush(errorResponse);
+      f.addListener(ChannelFutureListener.CLOSE);
```

Comment:
what does this listener do?

Reply:
it will wait for the thread to write the error response into the channel and flush, then close the channel.

Reply:
close the channel meaning close the conn? won't we reuse the channel?

Reply:
currently, we're not. I try to use the keep-alive header in HTTP protocol and don't close the channel. but most kinds of clients will hang there and wait for receiving continuing messages. I will try to fix it in recent days, perhaps they are in the next pr?